src/main/java/org/jenkinsci/plugins/workflow/support/steps/build/BuildTriggerListener.java

```diff
@@ -49,7 +50,8 @@ public void onCompleted(Run<?,?> run, @Nonnull TaskListener listener) {
                     trigger.context.onFailure(trigger.interruption);
                 }
             } else {
-                trigger.context.onFailure(new AbortException(run.getFullDisplayName() + " completed with status " + run.getResult() + " (propagate: false to ignore)"));
+                Result result = run.getResult();
+                trigger.context.onFailure(new FlowInterruptedException(result != null ? result : /* probably impossible */ Result.FAILURE, new DownstreamFailureCause(run)));
```

Comment:
One side effect of using `FlowInterruptedException` instead of `AbortException` is that `retry` will no longer retry the `build` step if it fails (it ignores `FlowInterruptedException` because of [JENKINS-44379](https://issues.jenkins-ci.org/browse/JENKINS-44379) ). This was reported as a bug in [JENKINS-60354](https://issues.jenkins-ci.org/browse/JENKINS-60354).

I'm not really sure how to fix this. I think `FlowInterruptedException` is overloaded with too many meanings at this point. One approach would be to factor out a `ResultCarryingException` interface, have `FlowInterruptedException` and some new `FooException` implement that interface, and then go through and switch everything that currently just uses `FlowInterruptedException` for a result to use `ResultCarryingException` when inspecting exception and `FooException` when creating them, and only use `FlowInterruptedException` when we only want to capture actual Pipeline interruption.

Reply:
I might rather suggest amending `RetryStep` to check for the presence of a `CauseOfInterruption`.

Reply:
> I might rather suggest amending RetryStep to check for the presence of a CauseOfInterruption.

Good idea, and much simpler to implement, although I do feel like it would be good to separate the different use cases of `FlowInterruptedException` at some point so that it is easier to reason about changes like this in isolation.

I think `catchError` and `warnError` probably have the same issue as `retry` and would need to be updated.

Reply:
Also, it looks like `WorkflowRun.doTerm` does not set a `CauseOfInterruption` when it aborts the build, so that would need to be tested and maybe changed since it _should_ be rethrown by `retry`. `WorkflowRun.doKill` does not set a `CauseOfInterruption` either, but I don't think it matters in that case because of the way the build is aborted.

Reply:
Perhaps just add a `fatal` flag to `FlowInterruptedException`, used by true build interruptions, and have steps like `retry` only ignore it in this mode. I guess that boils down to a similar proposal.